WIP: gpg: add pgp interface.

SantiagoTorres opened the pull request:
This adds the first stab at the signature generation and verification
routines. This is work in progress, as the codebase must be cleaned and
probably refactored into something more concise.

This is still a WIP branch, but we can start discussing and udpating the code
here.

SantiagoTorres committed:
WIP: gpg: add pgp interface.

This adds the first stab at the signature generation and verification
routines. This is work in progress, as the codebase must be cleaned and
probably refactored into something more concise.

SantiagoTorres commented on in_toto/gpg.py:
```diff
@@ -0,0 +1,247 @@
+    signature = gpg_sign_object(data)
+    print(signature)
+
+    pubkey = gpg_export_pubkey("903BAB73640EB6D65533EFF3468F122CE8162295")
```
this is my keyid, we still need to be clever-er at this, but this test code would work on your machine if you add the keyid of your default key in this line.

SantiagoTorres commented on in_toto/gpg.py:
```diff
@@ -0,0 +1,247 @@
+        raise Exception("This signature was truncated!")
+
+    return {
+        'keyid': "0x{}".format(binascii.hexlify(unhashed_subpackets).decode('ascii')),
```
we need a function to parse the unhashed_subpackets (get the keyid)

SantiagoTorres committed:
WIP: gpg: add python2 support

SantiagoTorres committed:
WIP: gpg: remove magic numbers

This should improve some readability.

SantiagoTorres committed:
WIP/STY: gpg: remove trailing whitespaces

SantiagoTorres committed:
WIP: gpg: refactor packet header parsing functions

Commit:
Add `--use-gpg` flag to in-toto-run to sign links

Adds the command line flag `--use-gpg` to `in-toto-run` and adapts
runlib and Metablock's signing function accordingly.

If the flag is added, then the argument passed via the `--key`
option is not interpreted as path but as GPG keyid instead.

The boolean set by the flag (`use_gpg`) is passed down to the
metadata signing function, i.e. `models.Metablock.sign`, where it
controls whether `securesystemslib` or the newly added `in_toto.gpg`
module is used for signing the Metadata.

SantiagoTorres committed:
WIP/ENH: add dsa signature verification

SantiagoTorres committed:
WIP/ENH: gpg: add test suite and test files

SantiagoTorres committed:
FIX: gpg: fix test-suite data location

SantiagoTorres committed:
FIX: metadata: fix gpg invocation

SantiagoTorres committed:
MAIN: gpg: force the use of gpgv2

Travis is running on ubuntu 14 and using gpg1 by default

SantiagoTorres committed:
ENH: test: change gpg keyring version to 2.1-

Tests fail on older systems due to new keyring secret keys being
unrecognizable if they are created on gnupg versions 2.1 and over.

SantiagoTorres committed:
FIX; gpg: fall back to default gpg , use sha256

lukpueh commented:
@SantiagoTorres, I hope you don't mind that I created a separate PR for the your neat editorconfig: #154

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,129 @@
+  if hash_algorithm not in SUPPORTED_HASH_ALGORITHMS:
+    raise Exception("This library only supports sha256 as the hash algorithm!")
+
+  # obtain the hased octets.
```
typo: hashed

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,129 @@
+    raise Exception("This signature algorithm is not supported!")
+  else:
+    key_type = SUPPORTED_SIGNATURE_ALGORITHMS[signature_algorithm]['type']
+    handler = SIGNATURE_HANDLERS[key_type]
```
You can remove the else clause here cf. [newly added recommendation about nesting](https://github.com/secure-systems-lab/code-style-guidelines#nesting)

lukpueh commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,112 @@
+      dsautils.Prehashed(hashing.SHA256())
+    )
+    return True
+  except:
```
I'd  `except cryptography.exceptions.InvalidSignature:` here

lukpueh commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,73 @@
+
+  keyarg = ""
+  if keyid:
+    keyarg="--default-key {}".format(keyid)
```
Please add whitespace

lukpueh commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,112 @@
+  See LICENSE for licensing information.
+
+<Purpose>
+  DSA-specific handling routines for signature verfication and key parsing
```
typo: verification

lukpueh commented on in_toto/gpg/rsa.py:
```diff
@@ -0,0 +1,81 @@
+  See LICENSE for licensing information.
+
+<Purpose>
+  RSA-specific handling routines for signature verfication and key parsing
```
typo: verification

lukpueh commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,135 @@
+    Parse an RFC4880 packet header and return its payload
+
+  <Arguments>
+    data: the packet header bufffer 
```
typo: buffer

lukpueh commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,135 @@
+    compute a keyid from an RFC4880 public-key buffer
+
+  <Arguments>
+    data: the public-key packet bufffer 
```
typo: buffer
arg name:`pubkey_packet_data`

lukpueh commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,135 @@
+  <Arguments>
+    data: the packet header bufffer 
+
+    expected_type:
```
What is expected type?

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,129 @@
+<Purpose>
+  signature-algorithm-agnostic frontend for the gpg_* parsing, signing and
+  verifying functions. These functions select the appropriate functions for
+  each algorithm and call it.
```
Looking at the module descriptions I wonder if it makes sense to to move the function `gpg_verify_signature` to  the module `gpg.functions`  and the functions `parse_pubkey_packet` and  `parse_signature_packet` to the module`gpg.util`. We wouldn't need `gpg.common` then. What do you think?

SantiagoTorres replied:
I mean that would work. I didn't give too much thought to the module layout. Should we go down that path?

lukpueh replied:
As you wish, this is really not a blocker.

lukpueh commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,135 @@
+    ptr = 2
+
+  if packet_type != expected_type:
+    raise Excption("This packet type is not supported!")
```
This should be an `Exception`, or better something like a custom `GpgException`.

lukpueh commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,135 @@
+    data: The MPI data
+
+  <Exceptions>
+    in-toto.gpg.TruncatedMPIException: if the data provided is not long enough
```
`TruncatedMPIException` does not exist yet. Maybe we should add a `gpg.exceptions` module?

SantiagoTorres replied:
Yup, this is something that I was going to add to the `__init__`

SantiagoTorres replied:
bump, what's our status here?

lukpueh replied:
Seems like it's still missing

lukpueh commented on in_toto/gpg/rsa.py:
```diff
@@ -0,0 +1,81 @@
+
+from in_toto.gpg.util import get_mpi_length, hash_object
+
+def create_key(pubkey_info):
```
Minor: Maybe call it `create_pubkey`. Same goes for `gpg.dsa.create_key`.

lukpueh reviewed (changes requested):
Looks pretty good to me.

I can't verify the validity of the key/signature parsing functions because I am not familiar with the standards but your unit tests are passing and look very reasonable.

I added a couple of minor inline comments that you might want to consider. Also could you also please add function docstrings to:
- `common.parse_pubkey_packet`
- `common.parse_signature_packet`
- `{rsa, dsa}.create_key`
- `{rsa, dsa}.get_pubkey_params`
- `{rsa, dsa}.get_signature_params`
- `{rsa, dsa}.gpg_verify_signature`
- `functions.gpg_sign_object`
- `functions.gpg_export_pubkey`

Another minor thing would be to remove trailing whitespace.

SantiagoTorres committed:
FIX: addresses review comments for the first round

Commit:
Add format schemas for in_toto.gpg keys

The module defines formats for RSA and DSA public key dicts
used for gpg signature verification.

SantiagoTorres committed:
ENH: gpg: add subpacket parsing routines

SantiagoTorres commented:
@lukpueh ugh, travis desn't generate long keyids because it's using an old version of gpg 👎

lukpueh commented:
We could disable in-toto's gpg signing/verifying support below a certain version of gpg.  Do you know the minimum version we would need?

Alternatively we could use short keyids, which as you mentioned is less secure (collision attacks).

Commit:
Replace dash in gpg signature field other-headers

Securesystemslib schema formats don't support dashes in dictionary
keys. To be able to use the existing infrastructure we change the
key's name from "other-headers" to "other_headers"

Commit:
Re-add keyid to gpg signature dictionary

336b34cf851cd85c7a049d6155c0fdba2ad20698 changed the way
the keyid is parsed from gpg signatures and accidentally removed
the keyid from the signature dict. This commit re-adds it.

Commit:
Remove use_gpg boolean from in_toto_run

Removes use_gpg boolean from in_toto_run (cli tool and runlib)
added in b08750348af51b09b08acff03e5606a1ec7a0b04.
I will add a separate gpg keyid parameter instead (see subsequent
commits)

This commit also slightly refactors `in_toto_run.in_toto_run` and
`runlib.in_toto_run` parameters. The `key` argument is now the last
argument and is called `signing_key`.
I do this because I will add additional key related arguments and
it seems neater to list them next to each other.

Commit:
Remove debug print from gpg.common

Commit:
Add util function to load public gpg keys as dict

This function can be used to e.g. add public keys to an in-toto
layout (trust model) or to load the layout signature verification
keys when running in-toto-verify.

Commit:
Add basic in-toto schema format module

This format uses securesystemslib schema verification and
currently defines formats for keys as specified in
securesystemslib.formats and in_toto.gpg.formats

Commit:
Fix double Metablock wrapping bug in in_toto_mock

in_toto_mock is using in_toto_run without key arguments and a
specific set of defaults, receiving an unsigned Metablock object
that contains a Link object from in_toto_run.

This commit removes wrapping the returned Metablock into another
Metablock inside in_toto_mock.

Commit:
Re-add gpg signing to in_toto_run and runlib

Adds new command line arguments to in-toto-run and arguments to
runlib.in_toto_run to alternatively sign a resulting link metadata
file with a gpg key identified by the passed gpg key id or with the
default gpg key in the keyring. Additionally a path to a gpg
keyring can be passed to not use the default gpg keyring.

This commit also adds a sign_gpg method to Metablock that uses the
new in_toto.gpg module instead of securesystemslib.

This commit also modifies both Metablock.sign and
Metablock.sign_gpg to return the newly created signature. This is
useful to e.g. obtain the used keyid, which is not available
beforehand when signing with the GPG default key.

Commit:
Add gpg support to final product verification

This commit adds support for verification of gpg signatures on
layout and link metadata.

The commit modifies in_toto_verify to accept additional parameters
to additionally load layout verification keys from the default or a
specified gpg keyring instead of key files as used by
securesystemslib.

The commit further modifies Metablock.verify_signatures to use
either in-toto.gpg or securesystemslib for signature verification
based on the signature type.

The commit also extends the allowed format for functionary public
keys on a layout (trustmodel), to allow adding public keys exported
from a gpg keyring.

Commit:
Add gpg support to in-toto-sign tool

Adds support for adding and verifying gpg signatures in
in-toto link or layout metadata.

Commit:
Add gpg to cli tool desiptions in README

Commit:
Make travis install gnupg2.1 from xenial repo

Commit:
Replace gpg with gpg2 command

Commit:
Remove gpg2 installation from .travis.xml

The unit tests still break, so this did not do the trick.

Commit:
Add gpg util function to get installed version

The added function uses the gpg2 command to get the version info
and parses the version string using regex.

Commit:
Add function to check if installed gpg2>=2.1.0

The function uses get_version to get the version of installed
gpg2 and compares it against a constant specifying the
minimal fully supported version number, needed to parse keyids
from signature packets.

Commit:
Warn if parse_signature_packet can't derive keyid

This commit changes parse_signature_packet to warn (instead of fail)
if no keyid can be derived from the signature data.

The signature dict is returned with an empty string as keyid and
the caller must deal with it appropriately.

Commit:
Add workaround if gpg signing can't compute keyid

This commit adds a workaround to gpg_sign_object if the created
signature is returned with an empty keyid.
The workaround consists of exporting the public part of the signing
key to compute the keyid.
Note: this will fail if gpg_sign_object is called without a keyid
(i.e. using the default key), because exporting the public key
requires a keyid.

Commit:
Adapt gpg tests to handle different gpg versions

Modifies the signing/verifying with default key functions to
expect an exception if the installed version of gpg is not
fully supported.

Adds additional tests to sign/verify with the keyid of the
signing key to the signing function, which should work on not
fully supported versions as well.

Also removes trailing whitespace.

Commit:
Omit test files and python packages in coverage

Commit:
Add two rsa keys to test gpg keyring

Commit:
Add keyid schema check to gpg_export_pubkey

Commit:
Raise exception when trying to parse empty pubkey

Commit:
Add gpg import pubkey test for in_toto.util

Commit:
Add gpg signing test for in-toto-run

Commit:
Add in-toto-verify tests for gpg signed supply chain

This commit copies the demo files used for testing where the
demo layout contains gpg public keys and the link files are signed
with the corresponding gpg private keys.

Commit:
Add gpg sign/verify test to in-toto-sign tool

Also adds "pragma: no branch" statements to `elif` blocks where
the default (`else`) is never reached, because of argument
validation in the argparser.

Note: the functions in in_toto_sign are only meant to be used with
the command line tool and therefor rely on the argparser's
validation.

Commit:
Return utf-8 encoded bytes as signable data

This commit changes Signable's signable_string property to return
utf-8 encoded bytes, which are expected by the signature creation
and verification methods.

Note that securesystemslib will re-encode the passed bytes, which
should be fixed for spec compliance.

The commit also re-signs the gpg test metadata files and fixes
comments.

Commit:
Add unittest for envvar set GNUPGHOME

Commit:
Exlcude gpg version case handling from coverage

This commit removes case handling for different versions of gpg
from code coverage (pragma: no cover) to not vary coverage in
different testing environments.
Once travis runs a newer system with an up-to-date version of gpg
when can remove this case handling altogether.

Commit:
Exclude gpg parsing errors from coverage

Crafting gpg signature and key data that triggers the excluded
exceptions doesn't seem feasible at the moment.

Commit:
Move gpg_verify_signature to gpg.functions module

The gpg interface module `functions` seems better suited for
the publicly available verification function.

Commit:
Add docstrings to gpg.functions

SantiagoTorres commented on .travis.yml:
```diff
@@ -12,7 +12,8 @@ install:
 
 script:
-    tox
+    - gpg2 --version
```
should we take this out of here?

lukpueh replied:
can do.

SantiagoTorres commented on README.md:
```diff
@@ -61,20 +61,32 @@ To learn more about the different rule types, their guarantees and how they are
+preliminary link file that only records the *materials*, then run the
+commands of that step, and finally use `in-toto-record ... stop ...` to
+record the *products* and generate the actual link metadata file. *Note:
+`in-toto-record` does not support GPG signing yet.*
```
worth adding a ticket for this.

lukpueh replied:
I don't know. Do we need it? I can also add it.

lukpueh replied:
I added this in [lukpueh/in-toto@add-gpg-in-toto-record](https://github.com/lukpueh/in-toto/tree/add-gpg-in-toto-record) and will submit a PR once this PR is merged.

SantiagoTorres commented on in_toto/in_toto_run.py:
```diff
@@ -104,7 +110,7 @@ def main():
   in_toto_args = parser.add_argument_group("in-toto options")
 
-  # FIXME: Do we limit the allowed characters for the name?
+   # FIXME: Do we limit the allowed characters for the name?
```
could you drop the stray space here?

lukpueh replied:
Can do

SantiagoTorres commented on in_toto/in_toto_run.py:
```diff
@@ -142,23 +155,44 @@ def main():
+    try:
+      key = util.prompt_import_rsa_key_from_file(args.key)
+
+    except Exception as e:
```
I think we want to catch the proper exception here.

lukpueh replied:
Could be any several Exceptions, ssl.FormatError, ssl.CryptoError, ssl.UnsupportedAlgorithmError, OSError, IOError ...

I thought catching a generic Exception is fine here because we want to exit in any case. Maybe something to re-think with https://github.com/in-toto/in-toto/issues/126?

What do you think?

SantiagoTorres commented on in_toto/in_toto_run.py:
```diff
@@ -142,23 +155,44 @@ def main():
-      parser.exit("For no command use --no-command option")
-    in_toto_run(args.step_name, args.materials, args.products,
-      args.link_cmd, key, args.record_streams)
+      parser.exit("For no command use `--no-command` option")
```
can we rephrase as follows:
```
ERROR: No command specified. Please specify (or use the --no-command option)
```
It feels a little bit clearer to me.

SantiagoTorres replied:
I'm wondering if refactoring this to
```python
if args.no_command:
  args.link_cmd = []
elif not args.link_cmd:
  parser.print_usage()
  parser.exit("...")

in_toto_run(...)
```
Would make this piece of code more readable.

lukpueh replied:
can do both

SantiagoTorres commented on in_toto/in_toto_sign.py:
```diff
@@ -84,25 +94,47 @@ def _sign_and_dump_metadata(metadata, args):
+    signature = None
+    # If the cli tool was called with `--gpg [KEYID ...]` `args.gpg` is
+    # a list (not None) and we will try to sign with gpg.
+    # If `--gpg-home` was not set, args.gpg_home is None and the signer tries
```
Just making sure: as of now it also respects if the `GNUPG_HOME` env_var is set before running

lukpueh replied:
yep that works

SantiagoTorres commented on in_toto/in_toto_sign.py:
```diff
@@ -84,25 +94,47 @@ def _sign_and_dump_metadata(metadata, args):
+
+      # Otherwise we sign with each passed keyid
+      for keyid in args.gpg:
+        # TODO: Add keyid schema check
```
Up [here](https://github.com/in-toto/in-toto/pull/150/files#diff-01f3f15afc9e2ce148a2237a8971e849R142) we are using `securesystemslib.schema.KEYID_SCHEMA`. Is this intended?

lukpueh replied:
I can add that check too

Commit:
Fix docstring and whitespace in gpg subpackage

Commit:
Remove print gpg version from travis config

Commit:
Add gpg id schema check to in-toto-sign

Commit:
Fix error msg and code readability in in-toto-run

Commit:
Adds signable_dict used with securesystemslib

In 8b2a9e65fd4befbd4a270808ccff3ab558b45363 we noted that
securesystemslib converts the received payload to canonical
JSON utf-8 encoded bytes before signing and verifying.

This commit adds dictionary property to Signable so that we can
easily pass the dictionary when using securesystemslib, to
prevent double encoding (for spec compliance).

The ideal fix would be to add functions to securesystemslib that
allow signing/verifying without re-encoding the payload.

vladimir-v-diaz commented on .coveragerc:
```diff
@@ -11,4 +11,6 @@ exclude_lines =
 
 omit =
+  */test/*
+  */python?.?/*
```
The commit message for this change says that `*/python?.?/*` are Python packages.  I haven't yet seen these folders in our projects.  Is this related to pyenv or virtualenv?

You should also verify whether some of the other omit lines are needed.  For example, I think `get_password` and `check_crypto_libraries` are no longer needed.

lukpueh replied:
`*/python?.?/*` excludes the parent directory of `*/site-packages/*` which also contains python modules, of which one made it into my coverage report recently, hence I added it to the omit section.

I'll remove the obsolete lines though.

vladimir-v-diaz commented on README.md:
```diff
@@ -61,20 +61,32 @@ To learn more about the different rule types, their guarantees and how they are
 
 ##### in-toto-run
-`in-toto-run` executes the passed command and records the path and hash of the passed *materials* - files before command execution - and *products* - files after command execution and optionally stores them together with the command's *byproducts* (e.g: return value, stdout or stderr) to a link file (`<step-name>.link`), signed with the functionary's key.
+`in-toto-run` executes the passed command and records the path and hash of
```
I found the use of the word "passed" here slightly ambiguous.  Is "given" a more effective alternative?
> in-toto-run executes the **given** command ...

It's unclear what is considered a command.   Some users might think it's a rule (e.g., in-toto-run CREATE `<pattern>`) or one of the command-line options (e.g., in-toto-run --key `<path>`), or some yet unspecified, supported command (e.g., in-toto-run start []), or some other application (e.g., in-toto-run tuf-repo --init).

lukpueh replied:
See issue #159

alyptik replied:
Maybe "executes the **specified** command..." would be a slightly clearer wording here.

vladimir-v-diaz commented on README.md:
```diff
@@ -61,20 +61,32 @@ To learn more about the different rule types, their guarantees and how they are
 
 ##### in-toto-run
-`in-toto-run` executes the passed command and records the path and hash of the passed *materials* - files before command execution - and *products* - files after command execution and optionally stores them together with the command's *byproducts* (e.g: return value, stdout or stderr) to a link file (`<step-name>.link`), signed with the functionary's key.
+`in-toto-run` executes the passed command and records the path and hash of
```
This paragraph might need a revision.  It's one long sentence and I find it difficult to determine its main purpose.

Can you state its main purpose in a shorter sentence?  For example,  "in-toto-run generates link metadata for the given command."

lukpueh replied:
See issue #159

vladimir-v-diaz commented on README.md:
```diff
@@ -61,20 +61,32 @@ To learn more about the different rule types, their guarantees and how they are
             [--products <filepath>[ <filepath> ...]]
             [--record-streams]
+            [--no-command]
             [--verbose] -- <cmd> [args]
```
The README places too much emphasis on the supported command-line options, which the user can easily learn about with `$ in-toto-run --help`.

I think it should instead include some super simple usage examples.  I see that it does link to a more in-depth example, but consider linking to it earlier in the README (perhaps after the simple usage example).

lukpueh replied:
See issue #159

vladimir-v-diaz commented on README.md:
```diff
@@ -61,20 +61,32 @@ To learn more about the different rule types, their guarantees and how they are
 
 ##### in-toto-record
-`in-toto-record` works similar to `in-toto-run` but can be used for multi-part software supply chain steps, i.e. steps that are not carried out by a single command. Use `in-toto-record ... start ...` to create a preliminary link file that only records the *materials*, then run the commands of that step, and finally use `in-toto-record ... stop ...` to record the *products* and generate the actual ...
+`in-toto-record` works similar to `in-toto-run` but can be used for
```
Do you think it's a good idea to make the main command `in-toto`?

```
$ in-toto <cmd>
```

Apart from being easier to type and remember, I feel users would instinctively attempt to execute `$ in-toto` after installing it.

 As for the other two supported commands, `in-toto-record` and `in-toto-verify`, I think the latter is fine.  I'm not so sure about the former command.  What about `in-toto-step` or `in-toto-interactive`?

lukpueh replied:
See issue #159

vladimir-v-diaz commented on in_toto/formats.py:
```diff
@@ -0,0 +1,40 @@
+  See LICENSE for licensing information.
+
+<Purpose>
+  Format schemas for in-toto metadata, based on securesystemslib.schema.
```
You should consider explaining how these schemata are used.  You can use the [text from securesystemslib.formats](https://github.com/secure-systems-lab/securesystemslib/blob/6d8abbc3adbfea511b76251693e1d6d894aae659/securesystemslib/formats.py#L28-L47).

lukpueh replied:
Addressed in 94986df

vladimir-v-diaz commented on in_toto/gpg/__init__.py:
```diff
@@ -0,0 +1,2 @@
+class PacketParsingError(Exception):
```
Should these exception definitions go in `in_toto/exceptions.py` or `in-toto/gpg/exceptions.py`?

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+        SUPPORTED_HASH_ALGORITHMS, SIGNATURE_HANDLERS, FULL_KEYID_SUBPACKET,
+        FULLY_SUPPORTED_MIN_VERSION)
+
+from in_toto.gpg.util import (compute_keyid, parse_packet_header,
```
> It is preferable to import an entire module than items from the module

https://github.com/secure-systems-lab/code-style-guidelines#imports

vladimir-v-diaz commented on in_toto/gpg/constants.py:
```diff
@@ -0,0 +1,56 @@
+"""
+<Module Name>
+  gpg/constants.py
```
We usually omit the full path, and just use the filename: `constants.py`.  The format of the module name should be consistent with how its done in the other modules.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+
+
+
+# XXX this doesn't support armored pubkey packets, so use with care.
```
How about linking to a page that explains armored pubkey packets?  Is the [OpenPGP Message Format RFC](https://tools.ietf.org/html/rfc4880#section-6.2) a good one?

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+
+# XXX this doesn't support armored pubkey packets, so use with care.
+# pubkey packets are a little bit more complicated than the signature ones
+def parse_pubkey_packet(data):
```
A lot of these functions need docstrings, per our guidelines.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+<Purpose>
+  signature-algorithm-agnostic frontend for the gpg_* parsing, signing and
+  verifying functions. These functions select the appropriate functions for
+  each algorithm and call it.
```
I'm not sure the description is accurate.  The two functions in this module don't appear to call any `gpg_*` functions.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+  if not data:
+    raise ValueError("Could not parse empty pubkey packet.")
+
+  data = parse_packet_header(data, PACKET_TYPES['main_pubkey_packet'])
```
Should `parse_packet_header()` return an exception if `data` contain bad bytes?  It seems to assume that data contains everything needed.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+  return key_params, keyinfo
+
+# this takes the signature as created by pgp and turns it into a tuf-like
+# representation (to be used with gpg_sign_object)
```
Maybe include the module (functions.py) where `gpg_sign_object()` is defined.  I had to search for it.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+  signature_type = data[ptr]
+  ptr += 1
+  if signature_type != SIGNATURE_TYPE_CANONICAL: # pragma: no cover
+    raise ValueError("We can only use canonical signatures on in-toto")
```
The error can be improved by explaining what's meant by canonical signature (PKCSv1.5 with RSA?) and specifying the invalid `signature_type` in the exception message.

I recommend that you verify the same for other exception messages.

SantiagoTorres replied:
Noted, they were pretty bad/out-of-context. Hope 1b620a1bc86c9da04e047b039c73aae980e464c2 addresses it.

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+  ptr += 1
+
+  if signature_algorithm not in SUPPORTED_SIGNATURE_ALGORITHMS: # pragma: no cover
+    raise ValueError("This signature algorithm is not supported!")
```
I'm not sure why some of these have `pragma: no cover`.  It's not easy to test for invalid signature algorithms because it's validated elsewhere?

lukpueh replied:
As mentioned in the commit message it does not seem feasible at the moment to craft signatures that would trigger all of these exceptions (ab26902ba1ea4db43e5c377ae4d2e539cc5d4438).

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+    raise ValueError("We can only use canonical signatures on in-toto")
+
+  signature_algorithm = data[ptr]
+  ptr += 1
```
Feels like I'm reading C code with all these `ptr += X`.  :P

SantiagoTorres replied:
😰

vladimir-v-diaz commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,151 @@
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
```
Is it a good idea to explain somewhere, maybe in the docstrings of the GPG modules, why other libraries can't be used for some of the functionality implemented here?  Is some of this not available in 'cryptography', 'gnupg,' etc.?   Is it to be compatible with Debian packaging?

vladimir-v-diaz reviewed (commented):
My review of `common.py`.

vladimir-v-diaz commented on in_toto/gpg/constants.py:
```diff
@@ -0,0 +1,56 @@
+      "type": "dsa",
+      "method": "pgp+dsa-fips-180-2",
+      "handler": dsa
+  }
```
I think `common.py` has an exception message stating that `0x01` is the only supported signature algorithm.  Why is `0x11` listed here?

lukpueh replied:
Where did you see that exception message?

vladimir-v-diaz replied:
I think it was this comment:

https://github.com/SantiagoTorres/in-toto/blob/778d10690dba95d67886f9a546a8778e1342ae30/in_toto/gpg/common.py#L81

There are also instances of "canonical signature type."  I'm not sure most readers will know what is meant by that term.

SantiagoTorres replied:
Hope 1b620a1bc86c9da04e047b039c73aae980e464c2 addressed this one

vladimir-v-diaz commented on in_toto/gpg/constants.py:
```diff
@@ -0,0 +1,56 @@
+SUPPORTED_HASH_ALGORITHMS = {0x08}
+SIGNATURE_TYPE_CANONICAL = 0x00
+FULL_KEYID_SUBPACKET = 0x21
+PARTIAL_KEYID_SUBPACKET = 0x10
```
Can you explain in a comment what `0x08` represents?  SHA256?
And the other values?

You can link to the RFC if it's more convenient.

SantiagoTorres replied:
hope 1b620a1bc86c9da04e047b039c73aae980e464c2 addresses this one

vladimir-v-diaz reviewed (commented):
My review of `constants.py`.

vladimir-v-diaz commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,114 @@
+from in_toto.gpg.util import get_mpi_length, hash_object
+import in_toto.gpg
+
+def create_pubkey(pubkey_info):
```
These function definitions are missing docstrings, per our guidelines.

vladimir-v-diaz commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,114 @@
+  p = int(pubkey_info['keyval']['public']['p'], 16)
+  q = int(pubkey_info['keyval']['public']['q'], 16)
+  parameter_numbers = dsa.DSAParameterNumbers(p, q, g)
+  pubkey = dsa.DSAPublicNumbers(y, parameter_numbers).public_key(backends.default_backend())
```
Verify that lines are less than 80 characters long, per our guidelines.

vladimir-v-diaz commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,114 @@
+  ptr += 2
+  value_y = data[ptr:ptr + value_y_length]
+  if len(value_y) != value_y_length: # pragma: no cover
+    raise in_toto.gpg.PacketParsingError("This MPI has been truncated!")
```
There are about four different places in `data` that can result in a truncated MPI error.  Should the exception message note where in `data` the truncation was detected?

vladimir-v-diaz commented on in_toto/gpg/dsa.py:
```diff
@@ -0,0 +1,114 @@
+    "q": binascii.hexlify(group_order_q).decode("ascii"),
+  }
+
+def get_signature_params(data):
```
I'm not sure why `params` is included in the function name.  It returns muItiple signature parameters?  Can it be renamed to `get_signature()` or `parse_signature()`?

vladimir-v-diaz reviewed (commented):
My review of `dsa.py`.

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+
+  <Returns>
+    The created signature in the format: gpg.formats.SIGNATURE_SCHEMA.
+  """
```
My EYES!  There are actually docstrings in here. :)

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+  command = GPG_SIGN_COMMAND.format(keyarg=keyarg, homearg=homearg)
+  process = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE,
+      stdin=subprocess.PIPE, stderr=None)
+  signature_data, _ = process.communicate(content)
```
We recently added a guideline for "junk" variables.
https://github.com/secure-systems-lab/code-style-guidelines#ignored-values

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+            Path to the gpg keyring. If not passed the default keyring is used.
+
+  <Exceptions>
+    None.
```
Zero exceptions?  What happens if the GPG subprocess doesn't generate a signature for whatever reason.  Does this function return an empty signature?

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+
+  signature = parse_signature_packet(signature_data)
+
+  # On GPG < 2.1 we cannot derive the keyid from the signature data.
```
The reader might want to know why GPG < 2.1 cannot derive keyids.

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+def gpg_sign_object(content, keyid=None, homedir=None):
+  """
+  <Purpose>
+    Calls gpg command line utility to sign the passed content with the key
```
Should you clarify that `gpg2` is called?  Consider pasting the full GPG command, or stating that GPG_SIGN_COMMAND is called.

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+    True if signature verification passes, False otherwise.
+
+  """
+  handler = SIGNATURE_HANDLERS[pubkey_info['type']]
```
Should you verify that `pubkey_info` is a `PUBKEY_SCHEMA`?

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+  command = GPG_EXPORT_PUBKEY_COMMAND.format(keyid=keyid, homearg=homearg)
+  process = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE,
+      stdin=subprocess.PIPE, stderr=None)
+  key_packet, _ = process.communicate()
```
An instance of a  "junk" variable that should follow our guidelines.
https://github.com/secure-systems-lab/code-style-guidelines#ignored-values

vladimir-v-diaz commented on in_toto/gpg/functions.py:
```diff
@@ -0,0 +1,168 @@
+  <Purpose>
+    Calls gpg command line utility to export the gpg public key identified by
+    the passed keyid from the gpg keyring at the passed homedir in a format
+    suitable for in-toto.
```
You should specify the GPG export command used.

vladimir-v-diaz reviewed (commented):
My review of `functions.py`.

vladimir-v-diaz commented on in_toto/gpg/rsa.py:
```diff
@@ -0,0 +1,83 @@
+  ptr += 2
+  exponent_e = data[ptr:ptr + exponent_e_length]
+  if len(exponent_e) != exponent_e_length: # pragma: no cover
+    raise in_toto.gpg.PacketParsingError("This e MPI has been truncated!")
```
This exception message can almost be confused for a typo.  Should it use "exponent e" ?

vladimir-v-diaz commented on in_toto/gpg/rsa.py:
```diff
@@ -0,0 +1,83 @@
+  ptr += 2
+  signature = data[ptr:ptr + signature_length]
+  if len(signature) != signature_length: # pragma: no cover
+    raise in_toto.gpg.PacketParsingError("This signature was truncated!")
```
I wonder if it's clearer to state that the MPI length in `data` doesn't match the signature length.  Or something along those lines.

vladimir-v-diaz commented on in_toto/gpg/rsa.py:
```diff
@@ -0,0 +1,83 @@
+
+  return signature
+
+def gpg_verify_signature(signature_object, pubkey_info, content):
```
I think there is a schema for `pubkey_info.`

vladimir-v-diaz reviewed (commented):
My review of `rsa.py`.

vladimir-v-diaz commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,222 @@
+import cryptography.hazmat.primitives.hashes as hashing
+import in_toto.gpg
+
+def get_mpi_length(data):
```
Should this link to an RFC?

vladimir-v-diaz commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,222 @@
+  bitlength = int(struct.unpack(">H", data)[0])
+  return int((bitlength - 1)/8) + 1
+
+def hash_object(headers, algorithm, content):
```
These functions are only separated from one another with a single blank line.  More than one blank is easier on the eyes.

vladimir-v-diaz commented on in_toto/gpg/util.py:
```diff
@@ -0,0 +1,222 @@
+    length = subpacket_octets[ptr]
+    ptr += 1
+    if length > 192 and length < 255 : # pragma: no cover
+      length = ((length - 192 << 8) + (subpacket_octets[ptr] + 102))
```
I've seen a lot of these magic numbers now.  Consider using constants for these.

SantiagoTorres replied:
I was thinking of doing this, but the RFC doesn't name them. I'll add a comment mentioning where the numbers come from.

vladimir-v-diaz reviewed (commented):
My review of `util.py.`

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -72,13 +79,10 @@ def in_toto_run(step_name, material_list, product_list,
+         record_streams, signing_key, gpg_keyid, gpg_use_default, gpg_home)
+
   except Exception as e:
     log.error("in toto run - {}".format(e))
```
Is the exact name of the command `in-toto-run` ?  You should use that.

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -114,9 +120,16 @@ def main():
+      "(if set without argument, the default key is used)"))
+
+  parser.add_argument("--gpg-home", dest="gpg_home", type=str,
+      help="Path to GPG keyring (if not set the default keyring is used)")
```
The default keyring location?

lukpueh replied:
Hm. I think I prefer just keyring.

vladimir-v-diaz reviewed (commented):
My review of `in_toto_run.py`.

Hmm, I just realized the command name is `in_toto_run`.  The documentation/README uses hyphens instead of underscores.

vladimir-v-diaz commented on in_toto/in_toto_sign.py:
```diff
@@ -30,12 +30,16 @@
                [--verify]
+
+  # gpg keys
+  in-toto-sign [-h] -f FILE -g [GPG [GPG ...]] [-o OUTPUT] [-a] [-v]
```
I don't recall the README mentioning this command.

vladimir-v-diaz replied:
The Purpose entry explains that this command can be used to verify signatures.   What's the different between this verify signature feature and `in_toto_verify.py?`

lukpueh replied:
Re README: see issue #159

lukpueh replied:
`in-toto-verify` is the command line interface for the final product verification implemented in verifylib, i.e. the entire routine that verifies if a final product was produced according to an in-toto layout (c.f. [`verifylib.in_toto_verify`s](https://github.com/in-toto/in-toto/blob/develop/in_toto/verifylib.py#L1249-L1289) docstring).

`in-toto-sign` on the other hand is a command line tool that can *only* add/replace/verify signatures in in-toto Link or Layout metadata. The tool was originally intended to sign layouts  created by the layout web wizard, but also serves well to re-sign test and demo data, e.g. if metadata formats or signing routines change.

vladimir-v-diaz commented on in_toto/in_toto_sign.py:
```diff
@@ -51,6 +55,11 @@
+  in-toto-sign -f gpg.layout --gpg
+
+  # Verify layout with a gpg key found in keyring
+  in-toto-sign -f gpg.layout --gpg 3BF8135765A07E21BD12BF89A5627F6BF439F3C2
```
Sign layout and Verify layout are the same?  This is odd.

lukpueh replied:
Oh, I forgot the `--verify` parameter. Good eyes!

vladimir-v-diaz commented on in_toto/in_toto_sign.py:
```diff
@@ -211,24 +258,46 @@ def main():
-      parser.exit(2, "wrong arguments: Link metadata can only be signed by"
-          " one key")
+      parser.exit(2, "too many arguments: Link metadata can not be signed by"
+          " multiple keys")
```
Why can't multiple keys sign for link metadata?  Not fully supported yet?  Limitation?  Prohibited?  Please explain in the exception message.

lukpueh replied:
Addressed in 778d10690dba95d67886f9a546a8778e1342ae30

vladimir-v-diaz reviewed (commented):
My review of `in_toto_sign.py.`

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -33,20 +33,28 @@
-    calls   verifylib.in_toto_verify   and handles exceptions.
+    Loads the layout metadata as Metablock object (containing a Layout object)
+    and the signature verification keys from the passed paths and/or from
+    layout_gpg_keyids, calls verifylib.in_toto_verify
```
Does `in_toto_verify.py` automatically match keyids to public keys?
Is the user required to list keys and keyids in the correct order?
[KEY1, KEY2, KEY3, ...] [KEYID1, KEYID2, KEYID3, ...]

You should document the expected behavior somewhere.

vladimir-v-diaz replied:
`in_toto_verify.py` can exit with a return value of 0, 1, or 2.  Should you document this in `<Purpose>`?

vladimir-v-diaz reviewed (commented):
My review of `in_toto_verify.py.`

Commit:
Remove obsolete lines from .coveragerc

The removed lines were added when securesystemslib was still part
of the tests (submodule). As it is installed as external
dependency we don't need this explicit excludes anymore.

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -146,14 +148,7 @@ def _validate_readme(self):
-
-    for keyid, key in six.iteritems(self.keys):
-      securesystemslib.formats.PUBLIC_KEY_SCHEMA.check_match(key)
+    in_toto.formats.ANY_PUBKEY_DICT_SCHEMA.check_match(self.keys)
```
Maybe the comment should explain that check_match() raises an exception for invalid keys.

Commit:
Add/update docstring in formats modules

Adds usage description and fixes a doc bug in gpg.formats.

Commit:
Add gpg.exceptions module

Adds exceptions module for gpg related errors and moves the
custom exception defined in gpg.__init__ to the new module.

Commit:
Replace item with module imports in gpg package

For code style guideline compliance
https://github.com/secure-systems-lab/code-style-guidelines#imports

Commit:
Replace package path with module name in doc header

Following a suggestion from @vladimir-v-diaz and to be consistent
with how it is done in other modules.

Commit:
Modify gpg docstrings to indiciate used gpg cmd

Commit:
Rename junk variable `_` to `junk`

This was recently added to the code style guidelines
https://github.com/secure-systems-lab/code-style-guidelines#ignored-values

Commit:
Sanitze inputs in gpg.gpg_verify_signature

Add schema check for the passed pubkey and signature objects
using schemas defined in gpg.formats.

Commit:
Update exception message in in_toto_run.in_toto_run

Following @vladimir-v-diaz's suggestion this commit changes the
error log prefix from "in toto run" to "in-toto-run".
The commit also adds the exception type to the error log.

Commit:
Update in-toto-sign doc header and exceptions msg

Adds missing option to usage section in doc header and refines
exception message, giving a tentative explanation why signing
Link metadata with multiple keys or adding signatures is not
allowed.

vladimir-v-diaz commented on in_toto/models/metadata.py:
```diff
@@ -125,33 +127,70 @@ def _type(self):
-      (see models.common.Signable.__repr__) of the Link or Layout object
-      contained in the `signed` property with the passed key and appends the
-      created signature to `signatures`.
+      Signs the pretty printed canonical JSON representation of the Link or
```
Why is it in pretty print format?  Explain in the comment.

vladimir-v-diaz commented on in_toto/models/metadata.py:
```diff
@@ -125,33 +127,70 @@ def _type(self):
     securesystemslib.formats.KEY_SCHEMA.check_match(key)
 
-    signature = securesystemslib.keys.create_signature(key, self.signed.signable_string)
+    # We actually pass the dictionary representation of the data to sign
```
The `<Purpose>` field seems to say something else about the format.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -314,25 +314,33 @@ def in_toto_mock(name, link_cmd_args):
+    gpg_use_default=False, gpg_home=None):
   """
   <Purpose>
     Calls function to run command passed as link_cmd_args argument, storing
```
Calls function?  Which function?

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -347,20 +355,28 @@ def in_toto_run(name, material_list, product_list,
   <Side Effects>
-    Writes newly created link metadata file to disk using the filename scheme
-    from link.FILENAME_FORMAT
+    If a key parameter is passed for signing, the ewly created link metadata
```
"ewly" typo.  We can't have users express distaste when creating link metadata.

vladimir-v-diaz commented on in_toto/util.py:
```diff
@@ -5,6 +5,8 @@
 import getpass
```
Some of these imported modules go unused in `util.py.`  For example, `json` and `pickle.`

vladimir-v-diaz reviewed (commented):
The rest of my comments for the modules.  I only glanced at the unit tests and didn't see anything wrong.

Commit:
Enhance docstrings in in_toto_run cli tool

- Mention exit codes.
- Mention layout signature verification behavior
- Add gpg signed layout scenario in usage section

Commit:
Fix function docstrings in metadata.Metablock

Removes obsolete mentionings of "pretty printed" canonical JSON
and emphasizes on the different payload type (bytes vs. dictionary)
required by the sign/verify backends (i.e. securesystemslib and
gpg).

Commit:
Update docstrings in runlib

- Fix a typo
- Clarify which functions are called in runlib.in_toto_run

Commit:
Remove unused module imports in util.py

Also fixes a non-style-guide conformant line continuation
indentation.

Commit:
Add blanklines, remove ws, refine docs in gpg

Commit:
Add docstrings to gpg dsa/rsa verify functions

lukpueh commented:
@SantiagoTorres, could you add the missing docstrings for 
`gpg.common.parse_pubkey_packet`, `gpg.common.parse_signature_packet`, `gpg.{rsa, dsa}.get_pubkey_params`, `gpg.{rsa, dsa}.get_signature_params`?

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,152 @@
+  if version_number not in SUPPORTED_PUBKEY_PACKET_VERSIONS: # pragma: no cover
+    raise ValueError("This pubkey packet version is not supported!")
+
+  time_of_creation = struct.unpack(">I", data[ptr:ptr + 4])
```
This never gets used.

SantiagoTorres replied:
I left these values in case we wanted to have a `-v` option akin to

```bash
Signature created on: Mon Jan  8 12:47:57 EST 2018
Hashed subpackets:
    .....
```

I also don't know how much it hurts to deserialize them, as it's a data structure following that format. I could probably leave a comment describing this though (and probably group the unused values). What do you think?

lukpueh replied:
Comment sounds good

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,152 @@
+  ptr += 2
+  unhashed_subpackets = data[ptr:ptr+unhashed_octet_count]
+  unhashed_subpacket_info = in_toto.gpg.util.parse_subpackets(
+      unhashed_subpackets)
```
`unhashed_subpacked_info` never gets used

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,152 @@
+        "you need at least gpg version '{}'. Your version is '{}'".format(
+        FULLY_SUPPORTED_MIN_VERSION, in_toto.gpg.util.get_version()))
+
+  left_hash_bits = struct.unpack(">H", data[ptr:ptr+2])[0]
```
`left_hash_bits` never gets used

lukpueh commented on in_toto/gpg/common.py:
```diff
@@ -0,0 +1,152 @@
+  ptr += 2
+
+  # Notice the /8 at the end, this length is the bitlength, not the length of
+  # the data in bytes (as len reports it)
```
I think this comment is outdated. Are you by any chance referring to [`util.get_mpi_length`](https://github.com/SantiagoTorres/in-toto/blob/41c58355d79af018f1800d5d002479bd8d377f8d/in_toto/gpg/util.py#L49)?

SantiagoTorres replied:
yeah, that comment was outdated. I moved it around on dd13b81afe3031ca955ec98c4898c46733b916dc

lukpueh commented:
The conflicts are easy to fix. But I'd rather not rebase on develop now, since GitHub will mark all pending comments as addressed after a rebase (if I'm not mistaken). So let's wait until all work is done here.

SantiagoTorres committed:
ENH: gpg: update comments for signature type

The signature type constant is mis-named (was canonical, should be
binary). Likewise, there is no no description of what the signature type
field meant. Fix the constant name, and add a comment describing what
the field does.

SantiagoTorres committed:
DOC: gpg: constants: describe hash constant values

The hash constant values appear like almost a magic number. Specify
where the values come from and add a descriptive constant for SHA256.

SantiagoTorres committed:
DOC:gpg:util:describe subpacket length computation

The subpacket length computation is full of magic numbers and a somewhat
convoluted logic. Explain where in the RFC is this logic described.

SantiagoTorres committed:
DOC:gpg:common: Improve exception message clarity

Most of the exception messages in the gpg.common module where quite
ambiguous and missing context. Improve the messaages in the exceptions
raised throughout the module.

SantiagoTorres committed:
DOC:gpg:functions: update exception docstrings

The gpg_sign_object did not mention two possible exceptions to raise
when trying to sign an object. Mention those two exceptions and the
reasons as to why they may be raised in the docstrings.

SantiagoTorres committed:
DOC:gpg: move outdated comment common -> util

The division by 8 comment was outdated in common, as the code with that
division was moved to util. Move the comment to the right location.

SantiagoTorres commented:
@lukpueh @vladimir-v-diaz pong. Let me know if this takes care of it

SantiagoTorres committed:
ENH: gpg: add missing docstrings

Many functions in the gpg module are missing docstrings. Add docstrings
to the public-facing functions to ease their use.

lukpueh commented:
Thanks for the additions! Could you add a docstring to `gpg.dsa.get_signature_params` and either remove or comment out the unused variables in `gpg.common` as pointed out above? Otherwise I think it is good to merge.

SantiagoTorres committed:
ENH: gpg: comment unused variables.

Some variables are unused in the common module functions. These
variables were added for readability. Comment these variables out and
describe their usage in case they are needed in the future.

Commit:
Minor form fixes in GPG subpackage

- Remove trailing whitespace
- Start comments with uppercase
- Fix typos
- Replace XXX comments with NOTE (linter doesn't like XXX)

Commit:
Disable pylint import checks for distutils

Pylint complains about distutils import in virtualenv (tox):
https://github.com/PyCQA/pylint/issues/73
```
E: 22, 0: No name 'version' in module 'distutils' (no-name-in-module)
E: 22, 0: Unable to import 'distutils.version.StrictVersion' (import-error)
```

Add inline pylint disable for `no-name-in-module` and
`import-error` checks as workaround for this issue.

Commit:
Remove unused six import in models.layout

Commit:
Replace filter/lambda with for loop in gpg.common

Pylint complains about lambda functions, and suggests a list
comprehension instead.
```
************* Module in_toto.gpg.common
W:184,10: map/filter on lambda could be replaced by comprehension (deprecated-lambda)
```

Our code style guideline prefers simple for loops over both:
https://github.com/secure-systems-lab/code-style-guidelines#no-lambda-functions-or-lisp-esque-code

Performance should not be an issue in this case. So let's use a
for loop.

Commit:
Add missing unit tests for layout validators

Commit:
Add no branch pragmas where eligible

This commit adds no branch pramgas in places that can't be
reached by unittests because of checks that happen earlier
in the control flow.

lukpueh commented:
8c30ac3 and dce3f07 are unrelated coverage improvement commits (actual new unittest and `no branch` pragmas where necessary) to pass unittests, which fail now if we go below 99%.
